In [368]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [369]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [370]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [371]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [372]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
market_data_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", "price_change_percentage_7d",
                                                              "price_change_percentage_14d", "price_change_percentage_30d",
                                                              "price_change_percentage_60d", "price_change_percentage_200d",
                                                              "price_change_percentage_1y"]])
market_data_scaled[0:5] 

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [373]:
# Create a DataFrame with the scaled data
df_market_data_scaled=pd.DataFrame(market_data_scaled,columns=["price_change_percentage_24h", "price_change_percentage_7d",
                                                              "price_change_percentage_14d", "price_change_percentage_30d",
                                                              "price_change_percentage_60d", "price_change_percentage_200d",
                                                              "price_change_percentage_1y"])


# Copy the crypto names from the original data
df_market_data_scaled["coin_id"]=df_market_data.index


# Set the coinid column as index
df_market_data_scaled=df_market_data_scaled.set_index("coin_id")

# Display sample data
df_market_data_scaled.head(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [374]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))


In [375]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)


C:\Users\aumek\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [376]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow.head(6) 

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   65.302379
5  6   52.888518

In [377]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
original_elbow= df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** The best value for K are 4 or 5 because anything below 4 are going to be to spread apart (large enertia), which would make it difficult to analyze.Anything higher than 4 or 5 has too much similarities, very low enertia, and would be difficult to segregate the variance. We probably should run 4 and 5 and see which one provides a more define cluster.

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [405]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)


In [406]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled) 

KMeans(n_clusters=4, random_state=1)

In [407]:
# Predict the clusters to group the cryptocurrencies using the scaled data
cryptocurrencies_predict = model.predict(df_market_data_scaled)


# Print the resulting array of cluster values.
print(cryptocurrencies_predict)

[3 3 1 1 3 3 3 3 3 1 1 1 1 3 1 3 1 1 3 1 1 3 1 1 1 1 1 1 3 1 1 1 0 3 1 1 2
 1 1 1 1]


In [408]:
# Create a copy of the DataFrame
cryptocurrencies_predict_df = df_market_data_scaled.copy()
cryptocurrencies_predict_df.head(3)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                             
bitcoin                      0.508529                    0.493193   
ethereum                     0.185446                    0.934445   
tether                       0.021774                   -0.706337   

          price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                              
bitcoin                      0.772200                     0.235460   
ethereum                     0.558692                    -0.054341   
tether                      -0.021680                    -0.061030   

          price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                               
bitcoin                     -0.067495                     -0.355953   
ethereum                    -0.273483                     -0.115759   
tether                       0.008005                     -0.550247   

          price_change_percentage_1y  
coin_id                               
bitcoin                    -0.251637  
ethereum                   -0.199352  
tether                     -0.282061

In [409]:
# Add a new column to the DataFrame with the predicted clusters
cryptocurrencies_predict_df['cryptocurrencies_predict'] = cryptocurrencies_predict


# Display sample data
cryptocurrencies_predict_df.head(3)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                             
bitcoin                      0.508529                    0.493193   
ethereum                     0.185446                    0.934445   
tether                       0.021774                   -0.706337   

          price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                              
bitcoin                      0.772200                     0.235460   
ethereum                     0.558692                    -0.054341   
tether                      -0.021680                    -0.061030   

          price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                               
bitcoin                     -0.067495                     -0.355953   
ethereum                    -0.273483                     -0.115759   
tether                       0.008005                     -0.550247   

          price_change_percentage_1y  cryptocurrencies_predict  
coin_id                                                         
bitcoin                    -0.251637                         3  
ethereum                   -0.199352                         3  
tether                     -0.282061                         1

In [411]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

predict_plot_original= cryptocurrencies_predict_df.hvplot.scatter(
    x="price_change_percentage_24h", 
    y="price_change_percentage_7d", 
    by="cryptocurrencies_predict",
    hover_cols = ["coin_id"]
)

predict_plot_original


:NdOverlay   [cryptocurrencies_predict]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [412]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [413]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
cryptocurrencies_predict_pca = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame. 
cryptocurrencies_predict_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [414]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_ 

array([0.3719856 , 0.34700813, 0.17603793])

In [415]:
pca.explained_variance_ratio_.sum()*100 

89.5031657030984

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance for the 3 principal components is 89.5%

In [416]:
# Create a new DataFrame with the PCA data.

# Creating a DataFrame with the PCA data
df_market_data_pca= pd.DataFrame(
    cryptocurrencies_predict_pca,
    columns=["PCA1", "PCA2", "PCA3"]
)

# Copy the crypto names from the original data
df_market_data_pca["coin_id"]=df_market_data.index

# Set the coinid column as index
df_market_data_pca=df_market_data_pca.set_index("coin_id")

# Display sample data
df_market_data_pca.head(5)

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [417]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [418]:
# Create an empty list to store the inertia values

inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_pca)
    inertia.append(k_model.inertia_)


C:\Users\aumek\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [419]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {"k": k, "inertia": inertia}
# Create a DataFrame with the data to plot the Elbow curve
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)
df_elbow_pca.head() 

k     inertia
0  1  256.874086
1  2  165.901994
2  3   93.774626
3  4   49.665497
4  5   37.878747

In [420]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
pca_elbow=df_elbow_pca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** the best K value under PCA is also 4.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** It does differ from the original as the inertia under the pca shows lower values (49 under pca  vs. 79 under original); therefore, the data points under the pca custering results will be closer to their cluster centroids, than the ones under the original data.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [421]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [422]:
# Fit the K-Means model using the PCA data
model.fit(df_market_data_pca) 

KMeans(n_clusters=4, random_state=1)

In [423]:
# Predict the clusters to group the cryptocurrencies using the PCA data
predict_pca = model.predict(df_market_data_pca)

# Print the resulting array of cluster values.
predict_pca

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [424]:
# Create a copy of the DataFrame with the PCA data
market_data_pca_predictions_df=df_market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
market_data_pca_predictions_df["predict_pca"] = predict_pca

# Display sample data
market_data_pca_predictions_df.head()

PCA1      PCA2      PCA3  predict_pca
coin_id                                                
bitcoin      -0.600667  0.842760  0.461595            1
ethereum     -0.458261  0.458466  0.952877            1
tether       -0.433070 -0.168126 -0.641752            0
ripple       -0.471835 -0.222660 -0.479053            0
bitcoin-cash -1.157800  2.041209  1.859715            1

In [425]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
predict_plot_pca= market_data_pca_predictions_df.hvplot.scatter(
    x="PCA1", 
    y="PCA2", 
    by="predict_pca",
    hover_cols=["coin_id"]
)
predict_plot_pca

:NdOverlay   [predict_pca]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [426]:
# Composite plot to contrast the Elbow curves
composite_elbow= original_elbow + pca_elbow
composite_elbow

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [428]:
# Composite plot to contrast the clusters
composite_plot= predict_plot_original + predict_plot_pca
composite_plot

:Layout
   .NdOverlay.I  :NdOverlay   [cryptocurrencies_predict]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [predict_pca]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Both, the original and pca elbow shows a value of 4; however, as mentioned earlier, pca elbow shows a lower value than the original providing a closer cluster, which is easier to analyze as we can see under PCA scareter plot above.